spring-kafka/src/main/java/org/springframework/kafka/support/serializer/JsonDeserializer.java

```diff
@@ -718,4 +717,27 @@ private JsonTypeResolver buildTypeResolver(String methodProperty) {
 		};
 	}
 
+	protected class CachedObjectReader {
```

Comment:
The problems are: 
- can't create ObjectWriter in ctor, because there's no way to provide customization of ObjectMapper in ctor
- can't create ObjectWriter in configure() function, because unit tests indicate it is not guaranteed that configure() will get called
- construct ObjectWriter at point of need, and possibly suffer performance issues

Therefore making no assumptions about how costly it is to construct an ObjectWriter.  Considering that the current serializer constructs the ObjectWriter just once in the constructor, I took the conservative approach and tried to at least cache the object so that if conditions are right it would still only get constructed once.  On the other hand, hashcode does have a cost, so if you don't think it matters to performance, I can pull it out.

Reply:
OK. Probably you just didn't see through my suggestion:

```
private volatile ObjectReader objectReader;
...
public T deserialize(...) {
    if (this.objectReader == null) {
		this.objectReader = this.objectMapper.readerFor(this.targetType);
	}
}
```

Does it make sense?
Technically your caching is just the same, but without `volatile`.
And that's why I don't see a reason in that extra instance if we just can deal with `volatile` and plain `if()`.

Sorry for not being clear enough from the beginning.

Reply:
That is certainly simpler, but I'm not familiar enough with all the supported code flows to decide if it's sufficient or not. 
 Are there any flows where a serializer might get reconfigured?  With this code, once the ObjectReader is constructed, changes to the ObjectMapper will not have any effect on the serialization (which was the original reason for the PR).  If that's an acceptable constraint, I'll make the change.

Reply:
I don't think so: you call `customizeObjectMapper()` from the `configure(Map<String, ?> configs, boolean isKey)` which is a Kafka Client API, which is called once from Kafka Client instance, e.g.:
```
 if (valueSerializer == null) {
                this.valueSerializer = config.getConfiguredInstance(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG,
                                                                                           Serializer.class);
                this.valueSerializer.configure(config.originals(Collections.singletonMap(ProducerConfig.CLIENT_ID_CONFIG, clientId)), false);
            } else {
                config.ignore(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG);
                this.valueSerializer = valueSerializer;
            }
```

And you see: they also set `config.ignore(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG);`, which means do not look into that property any more. So, any changes to the `ObjectMapper` at runtime are not going to be visible anyway.

Sounds reasonable?

Reply:
On the flip side, changes to the ObjectMapper may or may not be reflected in the hashcode.  If not, then the CachedObjectReader won't work any better (or worse) than a volatile.

Reply:
ok, will update.